src/Protocol/ProtocolRecognizer.cpp

```diff
@@ -1225,22 +222,132 @@ void cProtocolRecognizer::HandlePacketStatusRequest(void)
-	Pkt.WriteString(Response);
+	VERIFY(a_Out.WriteVarInt32(GetPacketID(cProtocol::ePacketType::pktStatusResponse)));
+	VERIFY(a_Out.WriteVarUTF8String(Response));
 }
 
 
 
 
 
-void cProtocolRecognizer::HandlePacketStatusPing()
+/* Ping handler for unrecognised versions. */
+static void HandlePacketStatusPing(cClientHandle & a_Client, cByteBuffer & a_Buffer, cByteBuffer & a_Out)
 {
 	Int64 Timestamp;
-	if (!m_Buffer.ReadBEInt64(Timestamp))
+	if (!a_Buffer.ReadBEInt64(Timestamp))
 	{
 		return;
 	}
 
-	cPacketizer Pkt(*this, pktPingResponse);
-	Pkt.WriteBEInt64(Timestamp);
+	VERIFY(a_Out.WriteVarInt32(GetPacketID(cProtocol::ePacketType::pktPingResponse)));
+	VERIFY(a_Out.WriteBEInt64(Timestamp));
+}
+
+
+
+
+
+std::unique_ptr<cProtocol> TryRecogniseProtocol(cClientHandle & a_Client, cByteBuffer & a_SeenData, std::string_view & a_Data)
+{
+	if (!a_SeenData.Write(a_Data.data(), a_Data.size()))
+	{
+		// We read more than the handshake packet here, oh well.
+
+		a_Client.Kick("Too much data was sent");
+		return {};
+	}
+
+	try
+	{
+		auto Protocol = TryRecognizeProtocol(a_Client, a_SeenData);
+		if (Protocol != nullptr)
+		{
+			// The protocol has just been recognized, advance data start
+			// to after the handshake and leave the rest to the protocol:
+			a_Data = a_Data.substr(a_SeenData.GetUsedSpace() - a_SeenData.GetReadableSpace());
+
+			// We read more than we can handle, purge the rest:
+#ifdef NDEBUG
+			a_SeenData.SkipRead(a_SeenData.GetReadableSpace());
+#else
+			ASSERT(a_SeenData.SkipRead(a_SeenData.GetReadableSpace()));
+#endif
+		}
+		else
+		{
+			a_SeenData.ResetRead();
+		}
+
+		return Protocol;
+	}
+	catch (const cUnsupportedButPingableProtocolException &)
+	{
+		// A server list ping for an unrecognised version is currently occuring
+		// Fall through to handler
+	}
+	catch (const std::exception & Oops)
+	{
+		a_Client.Kick(Oops.what());
+		return {};
+	}
+
+	cByteBuffer OutPacketBuffer(6 KiB);
+
+	// Handle server list ping packets
+	for (;;)
+	{
+		UInt32 PacketLen;
+		UInt32 PacketID;
+		if (
+			!a_SeenData.ReadVarInt32(PacketLen) ||
+			!a_SeenData.CanReadBytes(PacketLen) ||
+			!a_SeenData.ReadVarInt32(PacketID)
+		)
+		{
+			// Not enough data
+			break;
+		}
+
+		if ((PacketID == 0x00) && (PacketLen == 1))  // Request packet
+		{
+			HandlePacketStatusRequest(a_Client, OutPacketBuffer);
+		}
+		else if ((PacketID == 0x01) && (PacketLen == 9))  // Ping packet
+		{
+			HandlePacketStatusPing(a_Client, a_SeenData, OutPacketBuffer);
+		}
+		else
+		{
+			a_Client.Kick("Server list ping failed, unrecognized packet");
+			return {};
+		}
+	}
+
+	// Flush out all queued uncompressed, unencrypted data:
+	SendPacket(a_Client, OutPacketBuffer);
+
+	// In order to avoid making a whole dummy cProtocol just for old clients,
+	// we depend on cClientHandle calling us to process old pings. We don't
+	// persist state here, so restart parsing from scratch next time and return nothing:
+	a_SeenData.ResetRead();
```

Comment:
This seems like a weird workaround just to avoid making `cProtocolRecognizer` stateful. But I don't seem why it shouldn't be.